src/vcpkg/registries.cpp

```diff
@@ -1180,6 +1062,119 @@ namespace
 
 namespace vcpkg
 {
+    StringView VersionDbEntryDeserializer::type_name() const { return "a version database entry"; }
+    View<StringView> VersionDbEntryDeserializer::valid_fields() const
+    {
+        static const StringView u_git[] = {GIT_TREE};
+        static const StringView u_path[] = {PATH};
+        static const auto t_git = vcpkg::Util::Vectors::concat<StringView>(schemed_deserializer_fields(), u_git);
+        static const auto t_path = vcpkg::Util::Vectors::concat<StringView>(schemed_deserializer_fields(), u_path);
+
+        return type == VersionDbType::Git ? t_git : t_path;
+    }
+
+    Optional<VersionDbEntry> VersionDbEntryDeserializer::visit_object(Json::Reader& r, const Json::Object& obj)
+    {
+        VersionDbEntry ret;
+
+        auto schemed_version = visit_required_schemed_deserializer(type_name(), r, obj);
+        ret.scheme = schemed_version.scheme;
+        ret.version = std::move(schemed_version.versiont);
+
+        static Json::StringDeserializer git_tree_deserializer("a git object SHA");
+        static Json::StringDeserializer path_deserializer("a registry path");
+
+        switch (type)
+        {
+            case VersionDbType::Git:
+            {
+                r.required_object_field(type_name(), obj, GIT_TREE, ret.git_tree, git_tree_deserializer);
+                break;
+            }
+            case VersionDbType::Filesystem:
+            {
+                std::string path_res;
+                r.required_object_field(type_name(), obj, PATH, path_res, path_deserializer);
+                if (!Strings::starts_with(path_res, "$/"))
+                {
+                    r.add_generic_error(
+                        "a registry path",
+                        "A registry path must start with `$` to mean the registry root; e.g., `$/foo/bar`.");
+                    return nullopt;
+                }
+
+                if (Strings::contains(path_res, '\\'))
```

Comment:
I'm not happy about this breaking change for windows users, but since it's effectively UB between platforms I accept it.

Reply:
It's not a breaking change because the only practical way to author version database entries is with x-add-version or similar.

Reply:
We do not document x-add-version, but we have documented the exact format and syntax of the version database. Therefore, I consider the documentation to be binding and the behavior of `x-add-version` to not be.

Especially for filesystem registries, the ones affected by this change, I've personally witnessed users manually build them. `x-add-version` afaik does not support filesystem registries.

Reply:
>we have documented the exact format and syntax of the version database

I don't believe we actually did to that extent. What we do have always followed this form: https://github.com/microsoft/vcpkg/blob/master/docs/specifications/registries-2.md

![image](https://user-images.githubusercontent.com/1544943/127396776-c258327d-530d-4d51-82a4-f9507c6fa93d.png)

Also note that because stdfs::path did the splitting before, the behavior differed on Windows and Linux, which shouldn't be. Either we should always accept `\`, or never.

Reply:
Agreed, that's what I meant by "It was UB between platforms before" in my first message. Windows supported both, but linux would consider `\` part of the path element. It means that windows users which copy and pasted paths into, for example, vscode will be broken since the default behavior for VSCode upon pasting into a JSON document is to give something like

```json
  "key": "C:\\absolute\\path\\to\\my\\thing"
```
which the user may have trivially modified into
```json
  "key": "$\\to\\my\\thing"
```